python/cudf/cudf/tests/test_join_order.py

```diff
@@ -35,10 +41,124 @@ def right():
 # Behaviour in sort=False case didn't match documentation in many
 # cases prior to https://github.com/pandas-dev/pandas/pull/54611
 # (released as part of pandas 2.2)
-def expected(left, right, sort, *, how):
-    left = left.to_pandas()
-    right = right.to_pandas()
-    return left.merge(right, on="key", how=how, sort=sort)
+if PANDAS_GE_220:
+    # Behaviour in sort=False case didn't match documentation in many
+    # cases prior to https://github.com/pandas-dev/pandas/pull/54611
+    # (released as part of pandas 2.2)
+    def expected(left, right, sort, *, how):
+        left = left.to_pandas()
+        right = right.to_pandas()
+        return left.merge(right, on="key", how=how, sort=sort)
+
+else:
+
+    def expect_inner(left, right, sort):
```

Comment:
@galipremsagar @vyasr just confirming what we may have discussed offline a few months ago about testing older pandas versions, ideally we wouldn't want to keep around older-pandas-compatability-code like this and just xfail/skip a test in this case?

Reply:
Could you explain which tests you've chosen to xfail and which ones are skipped in this PR? AFAIK since all the `not PANDAS_GE_220` checks are negated, all the conditions are applying to tests that failed in older versions of pandas but pass on current pandas, which seems like they would be just as OK to skip. My comment above may have been unclear; to me, xfailing is most useful when we anticipate eventually getting an xpass as a signal. In these cases I don't think we ever will because we're xfailing tests on old versions that will only pass on newer versions. I'm fine using a different rationale if you think it's appropriate, but that's the source of my confusion here.

Reply:
I didn't originally add the skip vs xfail markers, but my guess it's was trying to adapt to a pattern where `request.applymarker` was used mid-test where `pytest.mark.skip` won't work and therefore `pytest.mark.xfail` was used instead.

I'll ensure we consistently use `skip` in the PR because you're correct that we shouldn't expect these old versions to every pass these tests

Reply:
I know we've come to some conclusion about when to `xfail` and when to `skip` but I don't recall. [I tend to favor `xfail`](https://github.com/rapidsai/cudf/pull/16595#discussion_r1725163751) as it gives us more information than `skip` (especially with strict xfail), so long as the test doesn't hang.

I would prefer that we use explicit version checks like "greater than 2.2.0" rather than "less than current version" but I hesitate to require that because it's a lot of work to verify and update, and I don't know if it's worth the time. Maybe we'll just delete the skip logic once we move to requiring pandas >= 3.

Reply:
> https://github.com/rapidsai/cudf/pull/16595#discussion_r1725163751 as it gives us more information than skip (especially with strict xfail), so long as the test doesn't hang.

I primarily prefer skips for tests that could never xpass, and therefore a skip is more appropriate. IOW a skip to me is for a test that will never work (which also to me means that a skip should always be a skipif, i.e. an unconditional skip is a meaningless test that can be deleted). An xfail is something that should pass, but doesn't because of a bug that might be fixed. The case we have here is a bit of both. We have behavior that we expect should pass but doesn't because of a bug in pandas, but we have also constrained the marker to only be applied to versions of pandas where it doesn't pass. Therefore, by definition we never expect a change because the versions of pandas that fail will always continue to fail, and only subsequent versions will pass. That to me feels more like a skip because the xfail will never become an xpass (on those failing versions). The xfail is masquerading as giving more information than it really does.

> I would prefer that we use explicit version checks like "greater than 2.2.0" rather than "less than current version" but I hesitate to require that because it's a lot of work to verify and update, and I don't know if it's worth the time.

We explicitly decided against that precisely because it's too much work to maintain. I can find and link you to the Slack threads where we discussed this if you'd like, but that's why the policy I linked above is the way it is.